### PR TITLE
Restructure code to be npm friendly, update build tools

### PR DESCRIPTION
**Pull request by lukescott**
This is a rather big PR. The functionality hasn't changed. Files were restructured and build tools were changed. 

The higher level changes are:

- Be more node friendly. Being able to require() a specific file and only include providers you care about. So if you're using webpack you can remove unused/unwanted code.
- Updated testing tools to something that has excellent coverage tools. Here is the current report for that:
```
-----------------|----------|----------|----------|----------|----------------|
File             |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-----------------|----------|----------|----------|----------|----------------|
All files        |    93.32 |    78.46 |      100 |    93.32 |                |
 src             |    98.78 |    93.02 |      100 |    98.78 |                |
  testUrls.js    |      100 |      100 |      100 |      100 |                |
  urlParser.js   |      100 |       92 |      100 |      100 |          21,34 |
  util.js        |    97.37 |    94.44 |      100 |    97.37 |             67 |
 src/provider    |    92.15 |    74.34 |      100 |    92.15 |                |
  canalplus.js   |    95.24 |       50 |      100 |    95.24 |             30 |
  coub.js        |    95.45 |       50 |      100 |    95.45 |             32 |
  dailymotion.js |    96.43 |       50 |      100 |    96.43 |             25 |
  template.js    |    92.86 |       50 |      100 |    92.86 |          36,41 |
  twitch.js      |    90.12 |    81.25 |      100 |    90.12 |... 101,121,122 |
  vimeo.js       |    90.63 |       50 |      100 |    90.63 |       31,32,56 |
  wistia.js      |    77.78 |    54.17 |      100 |    77.78 |... 71,79,80,92 |
  youku.js       |    96.97 |    66.67 |      100 |    96.97 |             42 |
  youtube.js     |     98.8 |    95.45 |      100 |     98.8 |             71 |
-----------------|----------|----------|----------|----------|----------------|
```
- Migrated from gulp to rollup for building bower dist file. npm doesn't need a build since it can use src/index.js or src/urlParser.js and src/providers/*.js directly.
- Updated from jshint to eslint (supports es6, and has a handy --fix option to fix most errors for you). I made sure to port the project's jshint config to eslint equivalent.

(More detailed list in commit message.)

Further changes I would like to make in the future:

- Get testing coverage up to 100%
- Further update code to es6 (classes, etc.)
- Ensure types are consistent. I noted some areas where sometimes strings or numbers were used. Will explore if there is a way to use TypeScript without actually converting the code to TypeScript files. Flow might be an option here, although in practice I found flow a bit annoying to use.
- Provide a TypeScript types definition file.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,34 +13,33 @@ Currently supports
...
+
+// Choose individual plugins
+import urlParser from "js-video-url-parser/lib/base";
+import "js-video-url-parser/lib/provider/canalplus';
+import "js-video-url-parser/lib/provider/coub';
+import "js-video-url-parser/lib/provider/dailymotion';
+import "js-video-url-parser/lib/provider/twitch';
+import "js-video-url-parser/lib/provider/vimeo';
+import "js-video-url-parser/lib/provider/wistia';
+import "js-video-url-parser/lib/provider/youku';
```
Comment (Zod-) [inline README.md]: Does this put the providers in the global namespace? I've had an issue before with a namespace clash (#25).

File: README.md [hunk 2]
```diff
@@ -13,34 +13,33 @@ Currently supports
...
-# grunt
-
-```shell
-$ grunt
-$ grunt build
-$ grunt test
-$ grunt dist
-
-#build and test specific plugins
-$ grunt --youtube --vimeo
```
Comment (Zod-) [inline README.md]: Is it still possible to build the dist for specific providers? People asked for it before so I'd like to keep it if possible. I'm fine with moving all the test stuff to npm but the grunt building part could be used in parallel?

File: README.md [hunk 3]
```diff
@@ -13,34 +13,33 @@ Currently supports
  - Coub
  - Wistia
 
-# grunt
-
-```shell
-$ grunt
```
Comment (Zod-) [inline README.md]: How do I build the dist locally?
```shell
npm
npm test
npm ?
```

Comment (lukescott) [reply README.md]: Update your local version of node to at least version 6. I'm on version 8. The code uses two ES2015 features: `const` and destructing.

npm should also be updated when you update node. npm updates your `package-lock.json` file automatically when you do npm commands. It lists versions that were actually installed so a build can be exactly reproduced. In contrast package.json lists package version ranges, which allows newer versions to be installed. Not to say that you can't install newer versions of dependencies, but package lock gives you a way to install exact versions. It helps with "but it works for me" situations. Very handy.

File: README.md (hunk 2 above)
Comment (lukescott) [reply README.md]: It is if you are using webpack, browserify, or something similar (anything that supports require - very common today). You import `js-video-url-parser/lib/base` first and then each plugin such as `js-video-url-parser/lib/youtube`. You don't need to build the project itself anymore. You simply include the parts you want and only those plugins get included during your own projects build process.

You only need a build for bower, which needs a dist file on GitHub that people can include with a script tag. I use rollup to produce this build. Most people today are using the npm registry though, and they actually don't need the produced build.

File: README.md (hunk 1 above)
Comment (lukescott) [reply README.md]: It does not. This is the ES2015 import syntax, similar to node's require. Each `.js` file is a separate isolated module.

**Comment by Zod-** [comment]
Tests are mostly lacking for the cases where the parser returns no result e.g.
```js
if (!result.id) { 
    return undefined; 
} 
```
So it shouldn't be too hard to get up to 100%.

I'd be fine with converting to classes/TypeScript. I thought about doing it myself at some point since I'm not happy with a lot of the parts, but I don't use the parser myself anymore so it's been tough to find motivation to work on it other then adding a requested site every few months.

**Pull request merged**